src/nb_func.cpp

```diff
@@ -184,6 +188,50 @@ char *strdup_check(const char *s) {
     return result;
 }
 
+#if defined(NB_FREE_THREADED)
+// Locked function wrapper for free-threaded Python
+struct locked_func {
```

Comment:
> make func_create in nb_func.h notice when it receives some arg_locked, arg_v_locked, or self_locked arguments, and conditionally instantiate a ft_object2_guard in its impl lambda (the most natural choice would be to use std::optional, but a custom RAII guard would also work if you don't want to take that header dependency)

This sounds really intriguing. What role does ``std::optional`` play? How would you conditionally insert the `ft_object2_guard` with templating, and how would it know which arguments to fetch? I had not considered this could be done in the compile-time part and don't see a reasonably simple solution.

Reply:
I’m on vacation until Sunday but will try to prototype this when I’m back!

Reply:
That would be amazing, thank you!

Reply:
One thing I wanted to add: I fear the meta-templating needed to pull this off might push things above a complexity threshold that I would be willing to accept. Right now, the solution has a small runtime cost, but users only need to pay that cost when using the locking feature. It's not perfect but an OK solution. Basically: don't spend a lot of time on it if you feel like the solution will turn into a template monstrosity.

Reply:
Take a look at https://github.com/wjakob/nanobind/pull/720 which implements this. I think it came out pretty reasonable as far as complexity goes.